File: engine/export/route.py
"""One route, on its own page: what is wrong, and which vehicles are carrying it.

WHY THIS EXISTS
===============
The board answers "which lanes are in trouble". It cannot answer the next
question a planner actually asks, which is "so what is on that lane, and which
of it is hurt" — and that question is about VEHICLES, not about lanes.

A lane in trouble is an abstraction. Nine trucks, of which two are stuck behind
a closed motorway and one has reported damage, is the thing somebody can act
on. So this view draws the route as its legs, and each leg as the vehicles
moving along it, each one coloured by its own state.

STRICTLY A PROJECTION
---------------------
Every figure here already appears on the planner's board. Nothing is
recomputed, and nothing is rounded differently. Two screens that disagree
about the same number destroy trust in both, and the one a planner will
believe is whichever they saw last — which is not a property you want.

THE THREE COLOURS
-----------------
``affected``  this leg is hit AND the option has closed, or somebody on site
              has reported damage. Red: it has already happened to this one.
``at_risk``   this leg is hit and there is still time. Amber: a decision.
``ok``        this leg is not hit. Green: leave it alone.

The distinction that matters is the middle one. A board that paints everything
touched in red tells a planner to panic about nine vehicles when two need a
decision today and the rest are fine for a week.
"""

from __future__ import annotations

from engine.export import progress as progress_mod
from engine.network.geo import Point, haversine_km
from engine.pipeline import RunContext


def _node_name(context: RunContext, node_id: str) -> str:
    node = context.config.nodes.get(node_id)
    return node.name if node else node_id


def _reports_for(shipment_id: str, context: RunContext) -> list[dict]:
    """Field reports on this consignment, as of the board's instant.

    The as-of filter is the point: a report that arrived after the instant
    this board describes must not appear on it, or the hindcast stops being
    reproducible the moment somebody files something.
    """
    from engine.ingest import reports as reports_mod  # noqa: PLC0415

    try:
        found = reports_mod.as_of(context.clock.as_of)
    except OSError:
        return []
    from engine.ingest import photos as photos_mod  # noqa: PLC0415

    out = []
    for report in found:
        if report.shipment_id != shipment_id:
            continue
        blob = report.as_dict()
        # The log stores ids; the rotation lives beside the file. Decorating
        # here keeps the append-only record free of anything derived — a log
        # that carries computed fields is a log that can disagree with the
        # thing it describes.
        blob["photos"] = [
            {"id": pid, "orientation": photos_mod.orientation_for(pid)}
            for pid in blob.get("photos", [])
        ]
        out.append(blob)
    return out


def route_view(board: dict, context: RunContext, route_id: str) -> dict | None:
    """The whole page, in one call."""
    route = next((r for r in board["routes"] if r["route_id"] == route_id), None)
    if route is None:
        return None

    shipments = [s for s in context.shipments if s.lane_id == route_id]
    hits_by_shipment: dict[str, set[int]] = {}
    for hit in context.hits:
        if hit.shipment_id in {s.shipment_id for s in shipments}:
            hits_by_shipment.setdefault(hit.shipment_id, set()).add(hit.leg_index)

    # Per-shipment figures, taken from the board rather than recomputed.
    risk_by_shipment: dict[str, dict] = {}
    for assessment in context.result.assessments:
        for risk in assessment.shipment_risks:
            current = risk_by_shipment.get(risk.shipment_id)
            lead = risk.lead_time_hours
            # Keep the SOONEST deadline across every event touching it: that
            # is the one that decides when somebody has to move.
            if lead is None:
                continue
            if current is None or lead < current["lead_time_hours"]:
                risk_by_shipment[risk.shipment_id] = {
                    "lead_time_hours": lead,
                    "actionability": getattr(risk, "actionability", None),
                    "expected_loss_chf": getattr(risk, "expected_loss_chf", 0.0),
                    "driving_event": assessment.event.title,
                    "driving_event_id": assessment.event.event_id,
                }

    reports_cache = {s.shipment_id: _reports_for(s.shipment_id, context) for s in shipments}

    # ---- the route drawn as legs, each carrying its vehicles -----------
    legs: list[dict] = []
    template = shipments[0].legs if shipments else []
    for index, leg in enumerate(template):
        vehicles = []
        for shipment in shipments:
            if index >= len(shipment.legs):
                continue        # a shipment routed short of the full lane
            own = shipment.legs[index]
            hit = index in hits_by_shipment.get(shipment.shipment_id, set())
            risk = risk_by_shipment.get(shipment.shipment_id)
            reports = reports_cache.get(shipment.shipment_id, [])
            damaged = any(r.get("load_state") == "damaged" for r in reports)
            stopped = any(r.get("status") in ("held", "stopped") for r in reports)

            if damaged or (hit and risk and risk["lead_time_hours"] <= 0):
                status = "affected"
            elif hit or stopped:
                status = "at_risk"
            else:
                status = "ok"

            moved = progress_mod.progress(shipment, context.config.nodes,
                                          context.clock.as_of)
            seen = progress_mod.observed(reports)
            vehicles.append({
                "progress": moved,
                "observed_position": seen,
                # How far the freight is from where the plan puts it. None,
                # not zero, when nothing has been observed — a zero would
                # read as "exactly on plan", which is the opposite of "we
                # have no idea where this is".
                "drift_km": progress_mod.drift_km(moved["planned_position"], seen),
                "shipment_id": shipment.shipment_id,
                "customer": shipment.customer,
                "carrier": own.carrier,
                "value_chf": shipment.value_chf,
                "status": status,
                "eta": shipment.eta.isoformat() if shipment.eta else None,
                "leg_arrives": own.planned_arrive.isoformat(),
                "lead_time_hours": (risk or {}).get("lead_time_hours"),
                "driving_event": (risk or {}).get("driving_event"),
                "expected_loss_chf": (risk or {}).get("expected_loss_chf", 0.0),
                "dangerous_goods": shipment.dangerous_goods,
                "temperature_controlled": shipment.temperature_controlled,
                "reports": reports,
            })

        counts = {
            "affected": sum(1 for v in vehicles if v["status"] == "affected"),
            "at_risk": sum(1 for v in vehicles if v["status"] == "at_risk"),
            "ok": sum(1 for v in vehicles if v["status"] == "ok"),
        }
        a_node = context.config.nodes.get(leg.from_node)
        b_node = context.config.nodes.get(leg.to_node)
        leg_km = (
            haversine_km(Point(a_node.lat, a_node.lon), Point(b_node.lat, b_node.lon))
            if a_node is not None and b_node is not None else 0.0
        )
        legs.append({
            "index": index,
            "km": round(leg_km, 1),
            "from": leg.from_node,
            "to": leg.to_node,
            "from_name": _node_name(context, leg.from_node),
            "to_name": _node_name(context, leg.to_node),
            "mode": leg.mode.value,
            "vehicles": vehicles,
            "counts": counts,
        })

    events = route.get("events") or []
    driving = max(
        events,
        key=lambda e: (e.get("shipments_here") or 0, e.get("exposure_chf") or 0),
        default=None,
    )

    return {
        "route_id": route_id,
        "name": route["name"],
        "level": route["level"],
        "level_label": route.get("level_label", route["level"]),
        "directive": route.get("directive", ""),
        "reason": route.get("reason", ""),
        "lead_time_hours": route.get("lead_time_hours"),
        "exposure_chf": route.get("exposure_chf"),
        "shipments_total": len(shipments),
        "shipments_affected": sum(
            1 for s in shipments if hits_by_shipment.get(s.shipment_id)
        ),
        "radar": route.get("radar"),
        # Both cuts, because the page draws both. Carried through rather than
        # recomputed: the board already did the split, and a second copy of
        # that arithmetic is a second thing to keep in step.
        "radar_measured": route.get("radar_measured"),
        "radar_reported": route.get("radar_reported"),
        "matrix_grid": board.get("matrix_grid"),
        "events": events,
        "driving_event_id": (driving or {}).get("event_id"),
        "legs": legs,
        "totals": {
            "affected": sum(leg["counts"]["affected"] for leg in legs),
            "at_risk": sum(leg["counts"]["at_risk"] for leg in legs),
            "ok": sum(leg["counts"]["ok"] for leg in legs),
        },
    }

File: api/static/route.js
/* One route, and the vehicles carrying it.
 *
 * The board says which lanes are in trouble. This page answers the question
 * that follows, and that question is about VEHICLES: nine trucks, of which
 * two are behind a closed motorway and one has reported damage, is something
 * a planner can act on. "A lane at Alert" is not.
 *
 * A PAGE, NOT A DIALOG. A planner looking at one lane wants to send somebody
 * the lane, and a modal cannot be sent. /route/LANE_RHINE_01 can.
 */
'use strict';

const $ = (id) => document.getElementById(id);
const esc = (s) => String(s ?? '').replace(/[&<>"]/g, (c) =>
  ({ '&': '&amp;', '<': '&lt;', '>': '&gt;', '"': '&quot;' }[c]));
const chf = (v) => v == null ? '—' : 'CHF ' + Math.round(v).toLocaleString('en-CH');

const ROUTE_ID = decodeURIComponent(location.pathname.split('/route/')[1] || '');
const state = { view: null, selected: null };

/* One glyph per mode. Drawn rather than emoji: an emoji renders differently
 * on every platform and at a size the browser picks, and these have to be
 * small, uniform and legible in a row of thirty. */
const GLYPH = {
  road:  '<rect x="1" y="6" width="12" height="8" rx="1.5"/><rect x="13" y="8" width="7" height="6" rx="1.5"/><circle cx="5" cy="16" r="2"/><circle cx="16" cy="16" r="2"/>',
  rail:  '<rect x="3" y="3" width="16" height="11" rx="2"/><circle cx="7" cy="16" r="1.8"/><circle cx="15" cy="16" r="1.8"/><path d="M1 19h20"/>',
  barge: '<path d="M2 13h18l-2 5H4z"/><rect x="6" y="6" width="10" height="6" rx="1"/>',
  sea:   '<path d="M2 13h18l-2 5H4z"/><rect x="6" y="5" width="10" height="7" rx="1"/><path d="M11 5V2"/>',
  air:   '<path d="M2 12l18-7-5 8 5 8-18-7z"/>',
};
const MODE_WORD = {
  road: 'trucks', rail: 'rail wagons', barge: 'barges',
  sea: 'vessels', air: 'aircraft',
};

function glyph(mode) {
  return GLYPH[mode] || GLYPH.road;
}

// ---------------------------------------------------------------
function statRow(v) {
  const cells = [
    ['ACTION BY', v.lead_time_hours == null ? '—'
      : (v.lead_time_hours < 48 ? `${Math.round(v.lead_time_hours)} h`
         : `${Math.round(v.lead_time_hours / 24)} days`),
      'until the first option closes'],
    ['EXPOSURE', chf(v.exposure_chf), `${v.shipments_affected} of ${v.shipments_total} consignments`],
    ['ALREADY HIT', String(v.totals.affected), 'no option left'],
    ['NEED A DECISION', String(v.totals.at_risk), 'still time to change them'],
  ];
  $('rt-stats').innerHTML = cells.map(([label, value, note]) => `
    <div class="rt-stat">
      <div class="rt-stat-label">${esc(label)}</div>
      <div class="rt-stat-value">${esc(value)}</div>
      <div class="rt-stat-note">${esc(note)}</div>
    </div>`).join('');
}

/* The convoy: one row per leg, one icon per consignment on it.
 *
 * Icons rather than a table because the question is "how many, and how bad",
 * which is a shape you read at a glance and a number you have to add up. */
function renderConvoy(v) {
  $('convoy').innerHTML = v.legs.map((leg) => {
    const word = MODE_WORD[leg.mode] || 'vehicles';
    const g = glyph(leg.mode);
    return `
    <div class="leg">
      <div class="leg-head">
        <div class="leg-route">
          <b>${esc(leg.from_name)}</b>
          <span class="leg-arrow">→</span>
          <b>${esc(leg.to_name)}</b>
        </div>
        <div class="leg-meta">
          ${leg.km ? `<span class="muted">${Math.round(leg.km)} km</span>` : ''}
          ${leg.vehicles.length} ${esc(word)}
          ${leg.counts.affected ? `<span class="pill pill--hit">${leg.counts.affected} hit</span>` : ''}
          ${leg.counts.at_risk ? `<span class="pill pill--risk">${leg.counts.at_risk} to decide</span>` : ''}
        </div>
      </div>
      <div class="leg-vehicles">
        ${leg.vehicles.map((veh) => `
          <button type="button" class="veh veh--${esc(veh.status)}"
                  data-leg="${leg.index}" data-ship="${esc(veh.shipment_id)}"
                  title="${esc(veh.shipment_id)} · ${esc(veh.customer)}">
            <svg viewBox="0 0 22 22" aria-hidden="true">${g}</svg>
            <span class="veh-id">${esc(veh.shipment_id.replace(/^SYN-/, ''))}</span>
          </button>`).join('')}
      </div>
    </div>`;
  }).join('');

  document.querySelectorAll('.veh').forEach((btn) => {
    btn.addEventListener('click', () => selectVehicle(btn.dataset.ship, +btn.dataset.leg));
  });
}

/* What one vehicle is carrying, and what somebody on site said about it.
 *
 * The reports are the reason this page is worth opening: every other number
 * here is inferred from a feed describing a region. A field report is
 * somebody looking at this consignment. */
function selectVehicle(shipmentId, legIndex) {
  const leg = state.view.legs.find((l) => l.index === legIndex);
  const veh = leg && leg.vehicles.find((x) => x.shipment_id === shipmentId);
  if (!veh) return;

  document.querySelectorAll('.veh').forEach((b) => b.classList.toggle(
    'is-on', b.dataset.ship === shipmentId && +b.dataset.leg === legIndex));

  const WORD = { affected: 'Already hit', at_risk: 'Needs a decision', ok: 'On plan' };
  const pr = veh.progress || {};
  const facts = [
    ['Customer', veh.customer],
    ['Carrier', veh.carrier],
    ['Value', chf(veh.value_chf)],
    ['On this leg by', new Date(veh.leg_arrives).toUTCString().slice(5, 22) + ' UTC'],
    ['Decide within', veh.lead_time_hours == null ? '—' : `${Math.round(veh.lead_time_hours)} h`],
    ['What is hitting it', veh.driving_event || 'nothing on this leg'],
  ];

  /* Distance done and distance left, with the bar showing the whole journey
   * rather than the current leg — "60% of the way to Rotterdam" is the shape
   * of the question, and a per-leg bar resets to zero at every node, which
   * reads as going backwards. */
  const journey = pr.total_km ? `
    <div class="prog">
      <div class="prog-bar"><span style="width:${Math.min(100, pr.percent)}%"></span></div>
      <div class="prog-nums">
        <b>${Math.round(pr.travelled_km).toLocaleString()} km</b> travelled
        <span class="muted">·</span>
        <b>${Math.round(pr.remaining_km).toLocaleString()} km</b> to go
        <span class="muted">of ${Math.round(pr.total_km).toLocaleString()} km (${pr.percent}%)</span>
      </div>
    </div>` : '';

  /* WHERE IT IS: two claims, never merged.
   *
   * The plan says one thing; somebody with the freight said another. A single
   * dot would have to pick, and whichever it picked it would be wrong half
   * the time — a planned dot is fiction the moment a truck stops, an observed
   * one is stale the moment it starts again. The GAP is the useful number. */
  const seen = veh.observed_position;
  const plan = pr.planned_position;
  const coord = (c) => `${c.lat.toFixed(3)}, ${c.lon.toFixed(3)}`;
  const mapLink = (c) =>
    `<a class="maplink" target="_blank" rel="noopener"
        href="https://www.openstreetmap.org/?mlat=${c.lat}&mlon=${c.lon}#map=11/${c.lat}/${c.lon}">open map</a>`;

  const where = `
    <div class="where">
      <div class="where-col">
        <span class="muted">WHERE THE PLAN PUTS IT</span>
        ${plan ? `<b>${coord(plan)}</b> ${mapLink(plan)}` : '<b>—</b>'}
        <em>interpolated along the leg; freight follows roads, not geodesics</em>
      </div>
      <div class="where-col">
        <span class="muted">LAST SEEN BY SOMEBODY</span>
        ${seen ? `<b>${coord(seen)}</b> ${mapLink(seen)}
           <em>${esc(seen.reported_by || 'on site')},
             ${esc(String(seen.observed_at).slice(0, 16).replace('T', ' '))} UTC${
               seen.accuracy_m ? ` · ±${Math.round(seen.accuracy_m)} m` : ''}${
               seen.first_hand === false ? ' · second hand' : ''}</em>`
          : '<b>not reported</b><em>nobody with this consignment has sent a position</em>'}
      </div>
      <div class="where-col">
        <span class="muted">OFF PLAN BY</span>
        ${veh.drift_km == null
          ? '<b>unknown</b><em>needs a position from site to compute</em>'
          : `<b class="${veh.drift_km > 50 ? 'drift-bad' : ''}">${veh.drift_km} km</b>
             <em>between where it should be and where it was seen</em>`}
      </div>
    </div>`;

  const reports = (veh.reports || []).length
    ? `<div class="rep-list">${veh.reports.map((r) => `
        <div class="rep">
          <div class="rep-top">
            <span class="tag ${r.load_state === 'damaged' ? 'tag--warn' : 'tag--ok'}">${esc(r.status)}</span>
            <b>${esc(r.role_label || 'on site')}</b>
            <span class="muted">${esc(String(r.observed_at).slice(0, 16).replace('T', ' '))} UTC</span>
            ${r.first_hand === false ? '<span class="tag tag--off">second hand</span>' : ''}
          </div>
          ${r.position ? `<div class="rep-line"><b>Where:</b> ${esc(r.position)}</div>` : ''}
          <div class="rep-line"><b>Load:</b> ${esc(r.load_state)}</div>
          ${r.note ? `<div class="rep-note">“${esc(r.note)}”</div>` : ''}
          ${(r.photos || []).length ? `<div class="rep-shots">${r.photos.map((ph) => {
            // EXIF is stripped on upload, so the rotation cannot come from
            // the file any more. It comes back as a number and is applied
            // here — otherwise stripping the metadata would quietly lay a
            // whole class of phone photos on their side.
            const id = typeof ph === 'string' ? ph : ph.id;
            const o = (typeof ph === 'object' && ph.orientation) || 1;
            return `<a href="/api/v1/photos/${esc(id)}" target="_blank" rel="noopener">
              <img class="shot-o${o}" src="/api/v1/photos/${esc(id)}"
                   alt="photo from site" loading="lazy">
            </a>`;
          }).join('')}</div>` : ''}
          ${r.lat != null ? `<div class="rep-line"><b>Fix:</b> ${r.lat.toFixed(4)}, ${r.lon.toFixed(4)}${r.accuracy_m ? ` ±${Math.round(r.accuracy_m)} m` : ''}</div>` : ''}
        </div>`).join('')}</div>`
    : `<p class="socket"><b>Nothing reported from the road.</b> Whoever is with
       this consignment can file in four taps at <code>/driver</code> — and a
       first-hand confirmation is what releases a re-route.</p>`;

  $('rt-vehicle-title').textContent = `${veh.shipment_id} — ${WORD[veh.status]}`;
  $('rt-vehicle').innerHTML = `
    ${journey}
    ${where}
    <div class="veh-facts">
      ${facts.map(([k, val]) => `
        <div><span class="muted">${esc(k)}</span><b>${esc(val)}</b></div>`).join('')}
    </div>
    <h3 class="rep-h">Reported from site</h3>
    ${reports}`;
  $('rt-vehicle-block').hidden = false;
  $('rt-vehicle-block').scrollIntoView({ behavior: 'smooth', block: 'nearest' });
}

// ---------------------------------------------------------------
async function boot() {
  const params = new URLSearchParams(location.search);
  const q = new URLSearchParams();
  if (params.get('as_of')) q.set('as_of', params.get('as_of'));
  if (params.get('shipments')) q.set('shipments', params.get('shipments'));

  const res = await fetch(`/api/route/${encodeURIComponent(ROUTE_ID)}?${q}`);
  if (!res.ok) {
    $('rt-name').textContent = 'That route is not on this board';
    return;
  }
  const v = await res.json();
  state.view = v;

  document.title = `${v.name} — Risk Radar`;
  $('rt-name').textContent = v.name;
  $('rt-chip').textContent = v.level_label || v.level;
  $('rt-chip').className = `level-chip level-${v.level}`;
  $('rt-directive').textContent = v.directive || '';

  statRow(v);
  renderConvoy(v);

  const onward = new URLSearchParams({ route: ROUTE_ID });
  if (params.get('as_of')) onward.set('as_of', params.get('as_of'));
  $('rt-next').href = `/ops?${onward}`;

  // Drawn by charts.js, the same code the board uses. A second copy of this
  // arithmetic would drift, and the day the two disagree is the day a planner
  // stops believing either.
  // Two panes. The legend is shared — the severity bands mean the same
  // thing on both — so it is drawn once, from whichever pane has data.
  const panes = [
    ['measured', v.radar_measured],
    ['reported', v.radar_reported],
  ];
  let legendDrawn = false;
  for (const [name, radar] of panes) {
    if (!radar) continue;
    Charts.drawRadar(radar, {
      svg: `rt-radar-${name}`,
      legend: legendDrawn ? null : 'rt-radar-legend',
      scale: 1.5,
    });
    if ((radar.max || 0) > 0) legendDrawn = true;
    const host = $(`rt-gauges-${name}`);
    if (host) {
      host.innerHTML = Charts.familyGauges(radar, {
        empty: name === 'measured'
          ? 'Nothing measurable is contributing delay here.'
          : 'Nothing reported is contributing delay here.',
      });
      Charts.wireGauges(host);
    }
  }
  const driving = (v.events || []).find((e) => e.event_id === v.driving_event_id)
               || (v.events || [])[0];
  $('rt-matrix').innerHTML = driving && driving.matrix
    ? Charts.buildMatrixGrid(driving, { matrix_grid: v.matrix_grid })
    : '<p class="muted">No event on this route carries a matrix.</p>';
}

boot();
